tests/torchtune/models/phi3/test_lora_phi3.py

```diff
@@ -0,0 +1,302 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from copy import deepcopy
+
+import pytest
+import torch
+
+from tests.test_utils import assert_expected, fixed_init_model
+from torch import nn
+from torchao.dtypes.nf4tensor import NF4Tensor
+from torchtune import utils
+from torchtune.models.phi3 import lora_phi3, phi3
+from torchtune.models.phi3._component_builders import lora_phi3_self_attention
+from torchtune.modules.peft import LoRALinear
+from torchtune.modules.peft.peft_utils import get_merged_lora_ckpt
+from torchtune.utils.seed import set_seed
+
+RANK = 4
+ALPHA = 1.0
+BSZ = 2
+SEQ_LEN = 32
+EMBED_DIM = 64
+INTER_DIM = 128
+NUM_HEADS = 4
+NUM_KV_HEADS = 2
+MAX_SEQ_LEN = 64
+
+
+@pytest.fixture(autouse=True)
+def random():
+    set_seed(16)
+
+
+class TestLoRAPhi3SelfAttention:
+    @pytest.fixture
+    def inputs(self) -> torch.Tensor:
+        inputs = torch.randn(BSZ, SEQ_LEN, EMBED_DIM)
+        return inputs
+
+    def get_lora_phi_self_attention(self, lora_modules):
+        lora_phi_sa = lora_phi3_self_attention(
+            lora_modules=lora_modules,
+            embed_dim=EMBED_DIM,
+            num_heads=NUM_HEADS,
+            num_kv_heads=NUM_KV_HEADS,
+            max_seq_len=MAX_SEQ_LEN,
+            lora_rank=RANK,
+            lora_alpha=ALPHA,
+        )
+        fixed_init_model(lora_phi_sa)
+        return lora_phi_sa
+
+    def test_empty_lora_modules(self):
+        with pytest.raises(ValueError, match="Must pass one or more of"):
+            _ = self.get_lora_phi_self_attention([])
+
+    @pytest.mark.parametrize(
+        "lora_modules, expected",
+        [
+            (["q_proj", "v_proj"], torch.tensor(50.64567)),
```

Comment:
where are these values generated from? Are we checking in scripts to reproduce these values?

Reply:
I don't believe we have a reference for phi3 LoRA. These values are just generated from the current code to check against future changes.